### PR TITLE
Update CONTRIBUTING.md to add guidance on supporting for .NET versions

### LINKED ISSUE #1031: Drop unsupported .NET runtime supports
The main repo has a clear statement about the .NET runtimes its supports : https://github.com/open-telemetry/opentelemetry-dotnet#supported-net-versions

The contrib repo does not have any such statement, and is mostly upto the individual components to state their supported runtimes. This is okay, given the components are contributed by different individuals/companies. However, there lacks any guidance for component authors regarding which ones they should (and should not) support. 

Proposing to add the following guidance (Draft)
1. Components SHOULD  support all officially supported .NET Runtimes.
2. Components SHOULD not support out-of-support .NET runtimes (eg: .NET 452, .NET Core 2.1 etc).
3. If a component chose to support out-of-support runtimes, the CI checks in this repo is *not* going to cover them. (TODO: Decide if we are moving to ci-per-component, to allow components to add unsupported .NET version to their  CI files).
4. Clearly document which runtimes are supported in the main readme.md.
5. Whenever a .NET version goes out of support, drop support for it in the component as well. (This does not require major version bump). (TODO: Add examples on when such changes were made previously).

What to do with existing components? 
There are few components right now which currently targets out-of-support versions. Will list them here and tag the owners so that they can start removing them. 
Once all components are done with the above, we'd need to remove CI checks for them (right now CI checks are done for .NET Core 3.1, which is an unsupported version)

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #1031 

## Changes

All the projects are now on supported version of .NET. This PR proposes guidance for new and existing component owners with respect to the .NET versions they should support.

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -258,3 +258,19 @@ the main branch. The workflow file should be named as
 [![NuGet version badge](https://img.shields.io/nuget/v/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 [![NuGet download count badge](https://img.shields.io/nuget/dt/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 ```
+
+### Guidance for target frameworks
+
+Components shipped from this repo SHOULD comply with the following requirements:
+
+* SHOULD support all the officially supported versions of
```
Comment (utpilla) [inline CONTRIBUTING.md:266]: Should we point to the supported versions of main repo instead and say that?

* SHOULD support the [.NET versions](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main#supported-net-versions) which are supported by the OpenTelemetry main repo packages.

Comment (vishweshbankwar) [reply CONTRIBUTING.md]: done

**Review by utpilla — APPROVED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -258,3 +258,18 @@ the main branch. The workflow file should be named as
...
+
+### Guidance for target frameworks
+
+Components shipped from this repo SHOULD comply with the following requirements:
+
+* SHOULD support the [.NET
+  versions](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main#supported-net-versions)
+  which are supported by the OpenTelemetry main repo packages.
+* SHOULD not support out-of-support .NET runtimes (eg: `.NET 452`, `.NET Core
+  2.1` etc). CI checks in this repository will not be run against out of support
```
Comment (cijothomas) [inline CONTRIBUTING.md:270]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1031#issuecomment-1450622575 Are we allowing/disallowing a component to define own CI ?

Comment (vishweshbankwar) [reply CONTRIBUTING.md:270]: Disallowing to define own CI for supporting out of support runtimes. At present we do not have any components that need this and for new components do not anticipate the need to support older versions.

**Review by Kielek — APPROVED** [review]

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -258,3 +258,18 @@ the main branch. The workflow file should be named as
 [![NuGet version badge](https://img.shields.io/nuget/v/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 [![NuGet download count badge](https://img.shields.io/nuget/dt/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 ```
+
+### Guidance for target frameworks
+
+Components shipped from this repo SHOULD comply with the following requirements:
+
+* MUST support the [.NET
```
Comment (cijothomas) [inline CONTRIBUTING.md:266]:
```suggestion
* SHOULD support the [.NET
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -258,3 +258,18 @@ the main branch. The workflow file should be named as
 [![NuGet version badge](https://img.shields.io/nuget/v/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 [![NuGet download count badge](https://img.shields.io/nuget/dt/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 ```
+
+### Guidance for target frameworks
+
+Components shipped from this repo SHOULD comply with the following requirements:
```
Comment (reyang) [inline CONTRIBUTING.md:264]: If `SHOULD` is used here, the bullet points don't need to and shouldn't have MUST/SHOULD anymore.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -258,3 +258,18 @@ the main branch. The workflow file should be named as
...
+
+* MUST support the [.NET
+  versions](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main#supported-net-versions)
+  which are supported by the OpenTelemetry main repo packages.
+* MUST not support out-of-support .NET runtimes (eg: `.NET 452`, `.NET Core
```
Comment (reyang) [inline CONTRIBUTING.md:269]:
```suggestion
* Not support out-of-support .NET runtimes (e.g.: `.NET Framework 4.5.2`, `.NET Core
```

Use the names from https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework.

File: CONTRIBUTING.md (hunk 3 above)
Comment (reyang) [reply CONTRIBUTING.md:266]:
```suggestion
* Support the [.NET
```

Comment (vishweshbankwar) [reply CONTRIBUTING.md]: updated

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -258,3 +258,16 @@ the main branch. The workflow file should be named as
 [![NuGet version badge](https://img.shields.io/nuget/v/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 [![NuGet download count badge](https://img.shields.io/nuget/dt/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 ```
+
+### Guidance for components on supporting target frameworks
+
+* Support the [.NET
```
Comment (cijothomas) [inline CONTRIBUTING.md:264]: looks like the usage of SHOULD and MUST is completely gone now?

Comment (vishweshbankwar) [reply CONTRIBUTING.md]: it was by mistake - fixed.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -258,3 +258,16 @@ the main branch. The workflow file should be named as
 [![NuGet version badge](https://img.shields.io/nuget/v/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 [![NuGet download count badge](https://img.shields.io/nuget/dt/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 ```
+
+### Guidance for components on supporting target frameworks
+
+* SHOULD support the [.NET
+  versions](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main#supported-net-versions)
+  which are supported by the OpenTelemetry main repo packages.
+* Must not support out-of-support .NET runtimes (e.g.: `.NET Framework 4.5.2`,
```
Comment (cijothomas) [inline CONTRIBUTING.md:267]:
```suggestion
* MUST NOT support out-of-support .NET runtimes (e.g.: `.NET Framework 4.5.2`,
```

**Review by cijothomas — APPROVED** [review]

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -258,3 +258,16 @@ the main branch. The workflow file should be named as
 [![NuGet version badge](https://img.shields.io/nuget/v/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 [![NuGet download count badge](https://img.shields.io/nuget/dt/{your_package_name})](https://www.nuget.org/packages/{your_package_name})
 ```
+
+### Guidance for components on supporting target frameworks
+
+* SHOULD support the [.NET
```
Comment (cijothomas) [inline CONTRIBUTING.md:264]: nit: add one more "MUST document the target frameworks supported in the main readme.md file".

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -258,3 +258,17 @@ the main branch. The workflow file should be named as
...
+  versions](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main#supported-net-versions)
+  which are supported by the OpenTelemetry main repo packages.
+* MUST document the target frameworks supported in the main readme.md file.
```
Comment (reyang) [inline CONTRIBUTING.md:267]:
```suggestion
* MUST document the target frameworks supported in the main README.md file.
```

**Review by reyang — APPROVED** [review]

**Comment by Kielek** [comment]
@alanwest, @CodeBlanch, could you please review?

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Review by alanwest — APPROVED** [review]

**Pull request merged**